Add Cloudflare malicious traffic runbook (#1858)
WordPress/openverse: documentation/meta/traffic/runbooks/identifying-and-blocking-traffic-anomalies.md

# Identifying and Blocking Traffic Anomalies in Cloudflare

## Objective

Enable consistent and prompt responses to traffic anomalies by implementing
procedures for identifying and blocking suspicious traffic—high request rates,
high error volume, or scraping patterns which disrupt service—in Openverse.

## When to use this runbook

> Use this runbook when either of these are true:
> 
> 1. There is a service disruption to the Openverse API or frontend.
> 2. There is a marked increase in resource usage on the API or frontend ECS
>    services that continues to trend to unsustainable levels, which _could lead_
>    to a service disruption.
> 
> And:
> 
> 1. There is a marked increase in traffic visible in our AWS dashboards or
>    Cloudflare analytics.

Reviewer: Would it be okay to put the necessary condition first? It's confusing to read in this order:

```suggestion
Use this runbook when there is a marked increase in traffic visible in our AWS dashboards or
   Cloudflare analytics and either of these are true:

1. There is a service disruption to the Openverse API or frontend.
2. There is a marked increase in resource usage on the API or frontend ECS
   services that continues to trend to unsustainable levels, which _could lead_
   to a service disruption.
```

3. Time. This is an exploratory process that can take significant trial-and
   error.

## Procedure Steps

### 1. Identify the timeframe of the traffic anomalies

1.  Log into Cloudflare.
2.  Select the relevant domain in the Cloudflare UI.
3.  Go to `Analytics & Logs => Traffic` and choose the "Page Views" box in the
    left sidebar.
4.  Set the time range dropdown to 24 hours.
5.  See if you can find the point where the current spike in traffic begins.
6.  If you do not see a place where traffic stabilizes to a relatively flat
    line, continue increasing the time range in the dropdown.
7.  Once you find where normalized traffic starts to spike in the graph, hover
    over the graph and **record the date and time the spike began**.

### 2. Identify targetable sources of the suspicious traffic

Next, we need to associate the traffic with specific sources. Essentially, we're
trying to find unique identifiers linked to potentially malicious traffic, so we
can block them through the Cloudflare Web Access firewall.

There are three primary identifiers for targeting traffic. These are ordered
from largest, most persistent, and most impactful, to smallest and most

> ephemeral. When possible, **blocking ASNs** is the single best way to mitigate

Reviewer: This is the first mention of ASN, so I'd suggest expanding it here first:

```suggestion
ephemeral. When possible, **blocking Autonomous System Numbers (ASNs)** is the single best way to mitigate
```

malicious traffic.

#### Traffic Source Identifiers

> - Source ASNs (Autonomous System Numbers): Unique identifiers for a group of IP

Reviewer: nit: usually it's clearer to write the abbreviation after the expanded version.

```suggestion
- Source "Autonomous System Numbers" (ASNs): Unique identifiers for a group of IP
```

  networks and routers under the control of a single organization, typically an
  internet service provider or hosting company.

>   Typically, a single ASN makes up less than 10% of total Openverse requests in
>   a given timeframe. If this percentage is larger, it could suggest that a user
>   on the ASN is scraping Openverse or accessing Openverse programmatically in an
>   aggressive way.

Reviewer: Slight rewording for clarity. The re-wording helps clarify for me what the difference between scraping and other programmatic usages are if they're both malicious.

```suggestion
  A single ASN usually makes up less than 10% of total Openverse requests in
  a given timeframe. If the percentage of requests from a single ASN is higher
  than 10%, it may indicate that one or more users on the ASN are scraping
  Openverse or over using/abusing expected programmatic access patterns.
```

Author: So much better, thanks!

- Source user agents: Unique user-agent strings sent with requests to Openverse.
  Nearly all UA strings from web browsers begin with `Mozilla`. Outside of the
  latest stable version of Chrome on Windows 10, which is the most popular
  Openverse browser (and usually looks something like
  `Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/{webkit-version} (KHTML, like Gecko) Chrome/{real-chrome-version} Safari/{webkit-version}`),

>   no single user agent should take up more than 10% of total traffic. Excluding

Reviewer: > no single user agent should take up more than 10% of total traffic

Won't this become less true over time as UA masking becomes more common? Where does this 10% number come from? Same question for the ASN 10% above as well. When would we know to adjust the expectation of "10%" specifically? Is it possible for us to create a dashboard that shows a rolling distribution of these so that we know what to benchmark against at any given moment?

Author: Creating a dashboard could be a good idea. For now I've added a note recommending users look at a time slice of traffic _without_ any known malicious traffic or incidents to compare the percentages.

Reviewer: Rewording it might be fine for now but it is still a bandaid until we have ongoing observability of what the expected rate should be. It probably won't be an issue for a little while still, UA masking isn't that aggressive yet, which is why it isn't a blocker to me.

  these with a `Source user agent does not contain Mozilla` or
  `Source browser equals Unknown` filter in the analytics UI can reveal other
  user agents from things like social media platforms and popular http request
  software libraries, like `Go-http-client/2.0` or `axios/0.21.4`, as examples.

- IP addresses: A unique identifier for a single machine or user.

  Typically, a single IP address makes up less than 1% of total Openverse
  requests in a given timeframe. These larger IPs typically belong to "good" web
  crawlers from places like Google, Bing, and Dotbot, as examples. If this
  percentage is larger, it could suggest that a user on the ASN is scraping

>   Openverse or accessing Openverse programmatically in an aggressive way.

Reviewer: Can you clarify whether the "larger IPs" go over 1%? We wouldn't want to unintentionally block search engine crawlers.

#### Steps

> 1.  Using the view from the previous task, scroll down and look for IP

Reviewer: Helpful cross-reference. The "previous task" is ambiguous and the section immediately preceding this one is non-trivial. It would be easy to get lost (I did!)

```suggestion
1.  Using [the view from the previous task](#1-identify-the-timeframe-of-the-traffic-anomalies), scroll down and look for IP
```

    Addresses, Source ASNs, or Source User Agents which exceed the typical usage
    thresholds defined above. Create a list of these traffic sources.

```{note}
Cloudflare will only show the 5 largest of each category; greater than 5 disproportionately-large items of any category would be unprecedented and surprising in Openverse's history.
```

```{warning}

> If you haven't found anything, it's unlikely that this situation is connected to malicious traffic or specific actors. Instead, it looks like a normal increase in traffic to Openverse. At this point, it's more likely that a change in code or infrastructure has caused a performance drop, exposed by the higher traffic, or that the traffic has grown so much that we need to allocate additional server resources.
> ```

Reviewer: ```suggestion
If you haven't found anything, it's unlikely that this situation is connected to malicious traffic or specific actors. Instead, it is more likely to be a normal increase in traffic to Openverse. At this point, it's more likely that a change in code or infrastructure has caused a performance drop, exposed by the higher traffic, or that the traffic has grown so much that we need to allocate additional server resources. Consider recent Openverse marketing efforts or WordPress events if the traffic looks organic and we need to attribute it to something.
```
```

2.  Take the list of suspicious traffic source IDs and go to the
    `Security => Bots` section of Cloudflare. Here, you'll see more lists of IP
    Addresses, ASNs, and User Agents below a graph of known bot traffic. It's
    very likely you'll see items from your lists in these graphs.

3.  Use the "add filter" button and filter by your listed resources
    individually. Write down which of the identified resources are comprised of
    the most `Automated` and `Likely Automated` traffic.

```{warning}
If an ASN has a significant amount of human traffic, it can indicate this ASN is an ISP rather than a hosting company or company which offers a Virtual Private Network. Blocking these types of ASNs can disrupt service to human users.
```

1. Finally, and optionally, make a web search in your preferred browser for each
   individual traffic source ID. Many ASNs and IP addresses used for scraping
   will have negative reports, be on spam lists, or be associated with hosted
   VPN services. This can be helpful in identifying which are truly malicious.
   However, our first priority is Openverse's availability to users, so this

>    step may only be necessary in lower-pressure scenarios.

Reviewer: Should we _always_ do this, but with the caveat that it can be done after the incident is resolved? It may help prevent long-term accidental or unintentional block against someone we do not want to block.

Author: > but with the caveat that it can be done after the incident is resolved? 

Great idea. Yeah, the reason to make it optional was that it can be time consuming, but doing it after is a great workaround. I'll update!

2.  Take the list of suspicious traffic source IDs and go to the
    `Security => Bots` section of Cloudflare. Here, you'll see more lists of IP
    Addresses, ASNs, and User Agents below a graph of known bot traffic. It's
    very likely you'll see items from your lists in these graphs.

3.  Use the "add filter" button and filter by your listed resources
    individually. Write down which of the identified resources are comprised of
    the most `Automated` and `Likely Automated` traffic.

```{warning}
If an ASN has a significant amount of human traffic, it can indicate this ASN is an ISP rather than a hosting company or company which offers a Virtual Private Network. Blocking these types of ASNs can disrupt service to human users.

> ````

Reviewer 2: This looks like it's not quite opened/closed properly, which is making the render look odd. Specifically (suggestions are hard WRT escaping the triple backticks) the first warning here is 4 backticks and doesn't appear to be closed at all.
![image](https://user-images.githubusercontent.com/10214785/235776617-33e25588-830e-42ca-ba63-f99eced635d4.png)

explain how to do this well. Some suggestions:

- Add the type of source being blocked to the name of the rule in brackets. For
  example, we currently use a `[bot-asn]` name prefix for all rules relating to
  ASNs.
- When writing rules which match ASNs, only the 5-digit number of the ASN is
  used in the rule. For this reason, it's nice to add the name of the ASN to the
  rule name as well. For example, rule name `[bot-asn] ZackWorld` which blocks
  ASN `13335`.
- Default to the "Block" action for each firewall rule.
- If a source id had a high percentage of human traffic and is making requests
  to the frontend, but you still feel it should be blocked, consider using the
  "Managed Challenge" action instead of an outright block. This should allow
  human users but block automated traffic.

### 4. Analyze reputation of identified traffic sources

Finally, make a web search in your preferred browser for each individual traffic
source identifier. Many ASNs and IP addresses used for scraping will have
negative reports, be on spam lists, or be associated with hosted VPN services.

> This can be helpful in identifying which are truly malicious.

Reviewer 2: Should this perhaps be step 3 instead? So that it comes before the instructions for blocking a traffic source.

Author: It actually was previously, but I moved it because it can be time consuming and I don't want it to block mitigating the issue. It's cool that this stood out to you though; I will add information explaining its positioning!

Reviewer 2: Ah okay, sounds good!

Periodically in the days after using this runbook repeat the entire process, at
your discretion, only continuing if new sources of traffic have appeared. When
blocking ASNs, it is unlikely that a scraping network, for example, would switch
regions or hosting companies. If new IPs with disproportionate traffic appear
and a similar traffic anomaly is present, this may suggest a botnet that hasn't
been appropriately targeted. Try to find an ASN or UA that matches the new IPs,
and can be correlated to the traffic from the original incident. Any targetable
property which they share in common could be a great block target.

## Escalation Procedures

1. If the team member is unable to complete the procedure successfully, escalate
   the issue to the rest of the team via a ping in our team chat. A maintainer
   should review your run of the book and repeat it if they find a particular
   type of traffic anomaly wasn't assessed.
2. If the team member is unable to complete the procedure successfully, escalate
   the issue to the rest of the team via a ping in our team chat. Here we will
   explore more drastic measures for mitigating the service disruption, like API

>    throttling or rate limiting adjustments.

Reviewer 2: These seem duplicated and/or irrelevant - do we just need the first point?
